### ISSUE TITLE
looking for help about https://git.zechert.net/fzechert/ansi-control-codes

### ISSUE THREAD

**Issue by orhanbalci** [fzechert/ansi-control-codes#1]
Hi @fzechert,
I'm really sorry to contact you from an unrelated repository. But could not find another way. I'm trying to use ansi-control-codes crate of yours to handle some tty parsing. Trying to convert [this](https://github.com/selectel/pyte/blob/master/pyte/streams.py) to rust. I have 2 questions;
1) Can you please export ascii macro so that one can generate non standart codes easily
2) can you provide a mechanism for control functions to easily compare with String and &str types. currently i convert control function to String and compare this with other string this seems a bit slugish performancewise
```rust
 if ESC.to_string() == char {
                    char = yield_!(None);
                    if char == "[" {
                        char = &CSI.to_string();
                    } else if char == "]" {
                        char = &OSC.to_string();
                    }
                }  
```

**Comment by fzechert** [comment]
I have moved the source code (and the issue) to github, so it becomes easier to discuss in the correct repository.

**Comment by fzechert** [comment]
### Regarding exporting the macro:

The macro is not public on purpose. As it invokes unsafe rust code, safety guarantee must come from the person that invokes the macro. For code inside this library, I - as a maintainer - can guarantee that calling code will not violate any safety concerns. If this macro would be public, this guarantee could not be made.
This macro is just a shorthand for some math / bit shifting on the arguments. I would advise you to just call 
```rust
unsafe { std::str::from_utf8_unchecked(&[xx]) }
``` 
with `xx` being your non-standard-code. Or of course you could just copy the macro into your own code, if you can make the necessary safety guarantees for all invocations.

### Regarding comparison between control functions and strings:

A control string (as the name suggests) can be longer than one character. Only the control codes defined in the module `c0` are single-character control functions. Comparing anything to a control string will always result in a string comparison, if we want a generic solution that works across all the modules (including `c1`, `control_sequences` / `control_strings` / `independent_control_functions`).

I could add some ease-of-use functions to the defined type (`ControlFunction`), but it would just hide the string conversion in a different place. So while this might be easier to use, it would not reduce complexity performance-wise.

You can already call
```rust
if String::from(ESC) == char {
  // ...
}
```
But this will do the same string conversion internally as your code already does.

The only alternative I see to string comparison is byte-array comparison. Each control function can be described by a sequence of bytes, that might be compared to another sequence of bytes. But this is just string comparison made on a lower level.

### In General

From what I see in your example code / python code, it looks like you are trying to **parse** control codes from some kind of input. This is not what this library was designed for. Maybe you will have an easier time with that task if you use a library that is more mature, or specifically designed for this purpose.

For example (not affiliated) https://docs.rs/crossterm/latest/crossterm/

**Comment by orhanbalci** [comment]
Thanks for your effort to move the repo here . Copied macro to my crate for the time being as you proposed.  For the comparison part it makes code more readable on my part if a comparison trait is implemented. Another thing is that since control functions own &str a conversion to String may not be necessary eleminating unnecessary allocations. Crossterm also is not suitable for parsing, it is oriented around composing such control strings easily. I really liked the idea how you compose codes using column/row notation.

**Comment by fzechert** [comment]
I am happy to implement something here that will help you. But it should be something that works across all the control functions.

As far as I am aware, the traits provided by [`std::cmp`](https://doc.rust-lang.org/std/cmp/index.html) are designed for comparison between to values of the **same** type. Since your goal is to compare between **different** types (e.g. String and ControlFunction, or &str and ControlFunction), those traits do not help with that goal.

From what I understand, there will always be some kind of conversion that needs to be invovled (`ControlFunction` to `&str`, or `ControlFunction` to `String`), then a comparison in a second step.

> since control functions own &str a conversion to String may not be necessary eleminating unnecessary allocations

I agree, but I do not (yet) see a good way to implement any kind of *free* conversion. Exposing a reference to this internal (e.g. by means of [`as_ref()`](https://doc.rust-lang.org/std/convert/trait.AsRef.html)) looks to me like something tha twould work for `c0`, but not for the other control characters.

The only way I see is to provide an implementation of [`PartialEq`](https://doc.rust-lang.org/std/cmp/trait.PartialEq.html) to do some inter-type comparisons.

If you have any good idea on this, feel free to share.

**Comment by orhanbalci** [comment]
I see. Either way it would be hacky solution. Is it possible to allow read only access for internal fields through getters? Then i can write my own comparison trait.

**Comment by fzechert** [comment]
Maybe have a look at this? https://github.com/fzechert/ansi-control-codes/pull/2/files

If this helps, I am glad to merge and release the changes.

This should enable you to do simple comparisons between string-like types (everything that implements `AsRef<str>`) and control functions.

For codes of the the modules `c0`, `c1`, and `independent_control_functions` this comes at almost no cost, at least without any allocation. For codes in module `control_sequences` it comes with the same cost as manually doing the `to_string()` conversion.

At least this should simplify your syntax and allow you to write stuff like this

```rust
let esc_control = c0::ESC;
let esc_str = "\u{001B}";

if esc_control == esc_str {
   // do something
}
```

**Comment by orhanbalci** [comment]
This is okay for me. For c0 codes i can directly use this for others i will update my parser logic to accumulate chars an then check for equality. Thanks for your help

**Comment by fzechert** [comment]
Published as `v0.0.4`.

You are welcome. If you have any more suggestions to improve the code, feel free to share.
